File: backend/app/services/drawings.py
import time

from fastapi import HTTPException, UploadFile
from sqlalchemy.orm import Session

from app.models.project import ProjectDrawing
from app.services.file_security import ALLOWED_DRAWING_EXTENSIONS, assert_allowed_extension, sanitize_path_segment
from app.services.storage import Storage
from app.services.zip_utils import ZipSecurityError, extract_zip, is_zip_filename


# Used by both project creation and "add more drawings" on an existing
# project. A zip is transparently extracted into one project_drawings row
# per file inside it; anything else is uploaded as-is. One bad file/zip
# entry doesn't block the rest — each is best-effort.
async def upload_drawings_for_project(
    db: Session, storage: Storage, project_id: str, files: list[UploadFile]
) -> dict[str, int]:
    uploaded = 0
    failed = 0

    for file in files:
        if not file or not file.filename:
            continue
        assert_allowed_extension(file.filename, ALLOWED_DRAWING_EXTENSIONS)
        content = await file.read()
        if not content:
            continue

        if is_zip_filename(file.filename, file.content_type):
            try:
                entries = extract_zip(content)
            except ZipSecurityError as exc:
                raise HTTPException(status_code=400, detail=str(exc))
            except Exception:
                failed += 1
                continue

            for entry in entries:
                path = f"{project_id}/{int(time.time() * 1000)}-{sanitize_path_segment(entry.name)}"
                try:
                    storage.save("project-drawings", path, entry.content, entry.content_type)
                except Exception:
                    failed += 1
                    continue
                db.add(ProjectDrawing(project_id=project_id, file_path=path, file_name=entry.name))
                uploaded += 1
            continue

        path = f"{project_id}/{int(time.time() * 1000)}-{sanitize_path_segment(file.filename)}"
        try:
            storage.save("project-drawings", path, content, file.content_type or "application/octet-stream")
        except Exception:
            failed += 1
            continue
        db.add(ProjectDrawing(project_id=project_id, file_path=path, file_name=file.filename))
        uploaded += 1

    db.commit()
    return {"uploaded": uploaded, "failed": failed}

File: src/lib/drawings.ts
import { SupabaseClient } from "@supabase/supabase-js";
import { isZipFile, extractZip } from "@/lib/zip";

// Used by both project creation and "add more drawings" on an existing
// project. A zip is transparently extracted into one project_drawings row
// per file inside it (so a contractor can browse/download individual
// drawings, not just one opaque archive); anything else is uploaded as-is.
// One bad file/zip entry doesn't block the rest — each is best-effort.
export async function uploadDrawingsForProject(
  supabase: SupabaseClient,
  projectId: string,
  files: File[]
): Promise<{ uploaded: number; failed: number }> {
  let uploaded = 0;
  let failed = 0;

  for (const file of files) {
    if (!file || file.size === 0) continue;

    if (isZipFile(file)) {
      let entries;
      try {
        entries = await extractZip(file);
      } catch {
        failed++;
        continue;
      }

      for (const entry of entries) {
        const path = `${projectId}/${Date.now()}-${sanitizePathSegment(entry.name)}`;
        const { error: uploadError } = await supabase.storage
          .from("project-drawings")
          .upload(path, entry.buffer, { contentType: entry.contentType });

        if (uploadError) {
          failed++;
          continue;
        }

        const { error: dbError } = await supabase
          .from("project_drawings")
          .insert({ project_id: projectId, file_path: path, file_name: entry.name });

        dbError ? failed++ : uploaded++;
      }
      continue;
    }

    const path = `${projectId}/${Date.now()}-${sanitizePathSegment(file.name)}`;
    const { error: uploadError } = await supabase.storage.from("project-drawings").upload(path, file);
    if (uploadError) {
      failed++;
      continue;
    }

    const { error: dbError } = await supabase
      .from("project_drawings")
      .insert({ project_id: projectId, file_path: path, file_name: file.name });

    dbError ? failed++ : uploaded++;
  }

  return { uploaded, failed };
}

// Storage paths can't safely contain arbitrary characters from a zip
// entry's internal path — keep the slashes (they become subfolders in
// the bucket, matching the zip's own folder structure) but strip
// anything else that could break a path segment.
function sanitizePathSegment(name: string): string {
  return name
    .split("/")
    .map((part) => part.replace(/[^a-zA-Z0-9._-]/g, "_"))
    .join("/");
}
